File: cua-helper/app/src/main/java/com/hermes/cua/CuaService.java
package com.hermes.cua;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.PixelFormat;
import android.hardware.display.DisplayManager;
import android.hardware.display.VirtualDisplay;
import android.media.Image;
import android.media.ImageReader;
import android.media.projection.MediaProjection;
import android.media.projection.MediaProjectionManager;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.util.DisplayMetrics;
import android.view.WindowManager;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.net.URLDecoder;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

public class CuaService extends Service {

    // Static fields for passing MediaProjection data from activity
    public static int pendingResultCode = -1;
    public static Intent pendingData = null;
    public static boolean isRunning = false;

    private static CuaAccessibilityService accService;

    public static void setAccService(CuaAccessibilityService s) {
        accService = s;
    }

    private MediaProjection mediaProjection;
    private VirtualDisplay virtualDisplay;
    private ImageReader imageReader;
    private int width, height, densityDpi;
    private ServerSocket serverSocket;
    private final AtomicBoolean running = new AtomicBoolean(false);
    private final Handler handler = new Handler(Looper.getMainLooper());

    @Override
    public void onCreate() {
        super.onCreate();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel ch = new NotificationChannel(
                    "cua", "CUA", NotificationManager.IMPORTANCE_LOW);
            ((NotificationManager) getSystemService(NOTIFICATION_SERVICE)).createNotificationChannel(ch);
            startForeground(1, new Notification.Builder(this, "cua")
                    .setContentTitle("Hermes CUA")
                    .setContentText("运行中 · :8640")
                    .setSmallIcon(android.R.drawable.ic_menu_manage)
                    .build());
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (!running.get()) {
            startHttp();
            isRunning = true;
        }

        if (pendingResultCode != -1 && pendingData != null) {
            try {
                setupProjection(pendingResultCode, pendingData);
            } catch (Exception e) {
                android.util.Log.e("CuaService", "setupProjection failed", e);
            }
            pendingResultCode = -1;
            pendingData = null;
        }
        return START_STICKY;
    }

    private void setupProjection(int resultCode, Intent data) {
        MediaProjectionManager mpm = (MediaProjectionManager) getSystemService(MEDIA_PROJECTION_SERVICE);
        if (mpm == null) { android.util.Log.e("CuaService", "mpm null"); return; }
        mediaProjection = mpm.getMediaProjection(resultCode, data);
        if (mediaProjection == null) { android.util.Log.e("CuaService", "mediaProjection null"); return; }

        WindowManager wm = (WindowManager) getSystemService(WINDOW_SERVICE);
        DisplayMetrics dm = new DisplayMetrics();
        try {
            wm.getDefaultDisplay().getRealMetrics(dm);
        } catch (Exception e) {
            wm.getDefaultDisplay().getMetrics(dm);
            android.util.Log.e("CuaService", "getRealMetrics failed", e);
        }
        width = dm.widthPixels;
        height = dm.heightPixels;
        densityDpi = dm.densityDpi;

        try {
            imageReader = ImageReader.newInstance(width, height, PixelFormat.RGBA_8888, 2);
        } catch (Exception e) {
            android.util.Log.e("CuaService", "ImageReader failed", e);
            return;
        }
        if (imageReader == null) {
            android.util.Log.e("CuaService", "imageReader null");
            return;
        }
        virtualDisplay = mediaProjection.createVirtualDisplay(
                "cua", width, height, densityDpi,
                DisplayManager.VIRTUAL_DISPLAY_FLAG_AUTO_MIRROR,
                imageReader.getSurface(), null, null);
    }

    private void startHttp() {
        running.set(true);
        new Thread(() -> {
            try {
                serverSocket = new ServerSocket(8640);
                serverSocket.setReuseAddress(true);
                serverSocket.setSoTimeout(2000);
                android.util.Log.i("CuaService", "HTTP server started on :8640");
                while (running.get()) {
                    Socket clientSocket = null;
                    try {
                        clientSocket = serverSocket.accept();
                        final Socket s = clientSocket;
                        // Handle each request in its own thread - prevents single stuck request blocking all others
                        new Thread(() -> handleRequest(s)).start();
                    } catch (SocketTimeoutException ste) {
                        if (clientSocket != null) try { clientSocket.close(); } catch (Exception ignored) {}
                    } catch (Exception e) {
                        android.util.Log.e("CuaService", "accept error: " + e.getMessage());
                        if (clientSocket != null) try { clientSocket.close(); } catch (Exception ignored) {}
                    }
                }
            } catch (Exception e) {
                android.util.Log.e("CuaService", "server died: " + e.getMessage());
            }
        }).start();
    }

    private void handleRequest(Socket s) {
        try {
            s.setSoTimeout(5000);
            InputStream in = s.getInputStream();
            java.io.BufferedReader reader = new java.io.BufferedReader(
                    new java.io.InputStreamReader(in, "UTF-8"));
            String requestLine = reader.readLine();
            if (requestLine == null || requestLine.isEmpty()) {
                s.close();
                return;
            }

            // Consume headers
            String header;
            while ((header = reader.readLine()) != null && !header.isEmpty()) {
                // consume
            }

            String[] parts = requestLine.split(" ", 3);
            String path = parts.length > 1 ? parts[1] : "/";
            String query = "";
            int qIdx = path.indexOf('?');
            if (qIdx >= 0) {
                query = path.substring(qIdx + 1);
                path = path.substring(0, qIdx);
            }

            OutputStream out = s.getOutputStream();
            boolean alreadyClosed = false;

            if (path.equals("/screenshot") || path.equals("/screen")) {
                serveScreenshot(out);
            } else if (path.equals("/ss2")) {
                // Debug: accessibility screenshot only - returns JSON
                if (accService == null) {
                    textResponse(out, "{\"error\":\"no accService\"}");
                } else {
                    byte[] png = accService.captureScreen();
                    textResponse(out, "{\"png\":" + (png == null ? "null" : ("" + png.length)) + ",\"sdk\":" + Build.VERSION.SDK_INT + ",\"status\":\"" + CuaAccessibilityService.lastCaptureError + "\"}");
                }
            } else if (path.equals("/dump")) {
                if (accService != null) {
                    final String[] dumpResult = new String[1];
                    final CountDownLatch dumpLatch = new CountDownLatch(1);
                    handler.post(() -> {
                        try { dumpResult[0] = accService.dumpUi(); }
                        catch (Exception e) { dumpResult[0] = "{\"error\":\"" + e.getMessage() + "\"}"; }
                        dumpLatch.countDown();
                    });
                    if (dumpLatch.await(2, TimeUnit.SECONDS)) {
                        textResponse(out, dumpResult[0]);
                    } else {
                        textResponse(out, "{\"error\":\"dump timeout\"}");
                    }
                } else textResponse(out, "{\"error\":\"no accessibility service\"}");
            } else if (path.equals("/status") || path.equals("/") || path.equals("/ping")) {
                textResponse(out, "OK");
            } else if (path.equals("/click")) {
                int x = getIntParam(query, "x", 0);
                int y = getIntParam(query, "y", 0);
                textResponse(out, "{\"success\":true}");
                out.flush();
                s.shutdownOutput();
                s.close();
                alreadyClosed = true;
                if (accService != null) {
                    final int fx = x, fy = y;
                    final CuaAccessibilityService as = accService;
                    new android.os.Handler(android.os.Looper.getMainLooper())
                        .postDelayed(() -> as.tap(fx, fy), 50);
                }
            } else if (path.equals("/swipe")) {
                int x1 = getIntParam(query, "x1", 0);
                int y1 = getIntParam(query, "y1", 0);
                int x2 = getIntParam(query, "x2", 0);
                int y2 = getIntParam(query, "y2", 0);
                int dur = getIntParam(query, "duration", 300);
                textResponse(out, "{\"success\":true}");
                out.flush();
                s.shutdownOutput();
                s.close();
                alreadyClosed = true;
                if (accService != null) {
                    final int fx1 = x1, fy1 = y1, fx2 = x2, fy2 = y2, fdur = dur;
                    final CuaAccessibilityService as = accService;
                    new android.os.Handler(android.os.Looper.getMainLooper())
                        .postDelayed(() -> as.swipe(fx1, fy1, fx2, fy2, fdur), 50);
                }
            } else if (path.equals("/back")) {
                textResponse(out, "{\"success\":true}");
                out.flush();
                s.shutdownOutput();
                s.close();
                alreadyClosed = true;
                if (accService != null) {
                    final CuaAccessibilityService as = accService;
                    new android.os.Handler(android.os.Looper.getMainLooper())
                        .postDelayed(() -> as.goBack(), 50);
                }
            } else if (path.equals("/home")) {
                textResponse(out, "{\"success\":true}");
                out.flush();
                s.shutdownOutput();
                s.close();
                alreadyClosed = true;
                if (accService != null) {
                    final CuaAccessibilityService as = accService;
                    new android.os.Handler(android.os.Looper.getMainLooper())
                        .postDelayed(() -> as.goHome(), 50);
                }
            } else if (path.equals("/input")) {
                String text = getParam(query, "text", "");
                if (accService != null && !text.isEmpty()) {
                    final String ft = text;
                    final CuaAccessibilityService as = accService;
                    textResponse(out, "{\"success\":true}");
                    out.flush();
                    s.shutdownOutput();
                    s.close();
                    alreadyClosed = true;
                    new android.os.Handler(android.os.Looper.getMainLooper())
                        .postDelayed(() -> as.inputText(ft), 50);
                } else {
                    textResponse(out, "{\"success\":false,\"error\":\"no text or service\"}");
                }
            } else if (path.equals("/openapp")) {
                String pkg = getParam(query, "pkg", "");
                if (pkg.isEmpty()) {
                    textResponse(out, "{\"success\":false,\"error\":\"missing pkg\"}");
                } else {
                    try {
                        Intent i = getPackageManager().getLaunchIntentForPackage(pkg);
                        if (i != null) {
                            i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK 
                                | Intent.FLAG_ACTIVITY_REORDER_TO_FRONT
                                | Intent.FLAG_ACTIVITY_CLEAR_TOP
                                | Intent.FLAG_ACTIVITY_SINGLE_TOP);
                            startActivity(i);
                            textResponse(out, "{\"success\":true}");
                        } else {
                            textResponse(out, "{\"success\":false,\"error\":\"app not found\"}");
                        }
                    } catch (Exception e) {
                        textResponse(out, "{\"success\":false,\"error\":\"" + e.getMessage() + "\"}");
                    }
                }
            } else if (path.equals("/copy")) {
                String text = getParam(query, "text", "");
                if (text.isEmpty()) {
                    textResponse(out, "{\"success\":false,\"error\":\"missing text\"}");
                } else {
                    ClipboardManager cm = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
                    cm.setPrimaryClip(ClipData.newPlainText("cua", text));
                    textResponse(out, "{\"success\":true}");
                }
            } else if (path.equals("/paste")) {
                if (accService != null) {
                    boolean ok = accService.pasteClipboard();
                    textResponse(out, "{\"success\":" + ok + "}");
                } else {
                    textResponse(out, "{\"success\":false,\"error\":\"no acc service\"}");
                }
            } else if (path.equals("/recents")) {
                if (accService != null) {
                    final CuaAccessibilityService as = accService;
                    textResponse(out, "{\"success\":true}");
                    out.flush();
                    s.shutdownOutput();
                    s.close();
                    alreadyClosed = true;
                    new android.os.Handler(android.os.Looper.getMainLooper())
                        .postDelayed(() -> as.performGlobalAction(android.accessibilityservice.AccessibilityService.GLOBAL_ACTION_RECENTS), 100);
                } else {
                    textResponse(out, "{\"success\":false,\"error\":\"no acc service\"}");
                }
            } else {
                textResponse(out, "404", 404);
            }
            if (!alreadyClosed) {
                out.flush();
                s.shutdownOutput();
                s.close();
            }
        } catch (SocketTimeoutException ste) {
            android.util.Log.w("CuaService", "request timeout");
            try { s.close(); } catch (Exception ignored) {}
        } catch (Exception e) {
            android.util.Log.e("CuaService", "handler: " + e.getMessage());
            try { s.close(); } catch (Exception ignored) {}
        }
    }

    private void serveScreenshot(OutputStream out) throws IOException {
        // Try MediaProjection first
        if (imageReader != null) {
            Image image = imageReader.acquireLatestImage();
            if (image != null) {
                Image.Plane[] planes = image.getPlanes();
                ByteBuffer buffer = planes[0].getBuffer();
                int ps = planes[0].getPixelStride();
                int rs = planes[0].getRowStride();
                int padding = rs - ps * width;
                Bitmap bitmap = Bitmap.createBitmap(width + padding / ps, height, Bitmap.Config.ARGB_8888);
                bitmap.copyPixelsFromBuffer(buffer);
                if (padding > 0) bitmap = Bitmap.createBitmap(bitmap, 0, 0, width, height);
                image.close();
                ByteArrayOutputStream baos = new ByteArrayOutputStream();
                bitmap.compress(Bitmap.CompressFormat.PNG, 90, baos);
                sendPng(out, baos.toByteArray());
                return;
            }
        }
        // Fallback: AccessibilityService screenshot (API 34+, works on HyperOS)
        if (accService != null && Build.VERSION.SDK_INT >= 34) {
            try {
                byte[] png = accService.captureScreen();
                if (png != null && png.length > 100) {
                    sendPng(out, png);
                    return;
                }
            } catch (Exception e) {
                android.util.Log.e("CuaService", "acc screenshot failed", e);
            }
        }
        textResponse(out, "503", 503);
    }

    private void sendPng(OutputStream out, byte[] bytes) throws IOException {
        StringBuilder headers = new StringBuilder();
        headers.append("HTTP/1.1 200 OK\r\n");
        headers.append("Content-Type: image/png\r\n");
        headers.append("Content-Length: ").append(bytes.length).append("\r\n");
        headers.append("Connection: close\r\n\r\n");
        out.write(headers.toString().getBytes());
        out.write(bytes);
        out.flush();
    }

    private void textResponse(OutputStream out, String text) throws IOException {
        textResponse(out, text, 200);
    }

    private void textResponse(OutputStream out, String text, int code) throws IOException {
        byte[] bytes = text.getBytes("UTF-8");
        StringBuilder headers = new StringBuilder();
        headers.append("HTTP/1.1 ").append(code).append(" ");
        if (code == 200) headers.append("OK");
        else if (code == 404) headers.append("Not Found");
        else if (code == 500) headers.append("Internal Error");
        else if (code == 503) headers.append("Service Unavailable");
        headers.append("\r\n");
        headers.append("Content-Type: text/plain; charset=UTF-8\r\n");
        headers.append("Content-Length: ").append(bytes.length).append("\r\n");
        headers.append("Connection: close\r\n\r\n");
        out.write(headers.toString().getBytes());
        out.write(bytes);
        out.flush();
    }

    private int getIntParam(String query, String key, int def) {
        try { return Integer.parseInt(getParam(query, key, String.valueOf(def))); }
        catch (Exception e) { return def; }
    }

    private String getParam(String query, String key, String def) {
        if (query == null) return def;
        for (String p : query.split("&")) {
            String[] kv = p.split("=", 2);
            if (kv.length == 2 && kv[0].equals(key))
                return URLDecoder.decode(kv[1]);
        }
        return def;
    }

    @Override
    public void onDestroy() {
        isRunning = false;
        running.set(false);
        try { if (serverSocket != null) serverSocket.close(); } catch (Exception ignored) {}
        try { if (virtualDisplay != null) virtualDisplay.release(); } catch (Exception ignored) {}
        try { if (imageReader != null) imageReader.close(); } catch (Exception ignored) {}
        try { if (mediaProjection != null) mediaProjection.stop(); } catch (Exception ignored) {}
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) { return null; }
}

File: cua-helper/app/src/main/java/com/hermes/cua/CuaService.kt
package com.hermes.cua

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.graphics.Bitmap
import android.hardware.display.DisplayManager
import android.hardware.display.VirtualDisplay
import android.media.ImageReader
import android.media.projection.MediaProjection
import android.media.projection.MediaProjectionManager
import android.os.Build
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.util.DisplayMetrics
import android.view.WindowManager
import android.content.Context
import fi.iki.elonen.NanoHTTPD
import java.io.ByteArrayOutputStream

class CuaService : Service() {

    private var httpServer: CuaHttpServer? = null
    private var mediaProjection: MediaProjection? = null
    private var virtualDisplay: VirtualDisplay? = null
    private var imageReader: ImageReader? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    companion object {
        const val PORT = 8765
        var mediaProjectionResult: Pair<Int, Intent>? = null
        var isRunning = false
        var mediaProjection: MediaProjection? = null
            private set

        // Store the last screenshot as bytes
        var lastScreenshotBytes: ByteArray? = null
    }

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val notification = createNotification()
        startForeground(1, notification)

        // Initialize MediaProjection
        mediaProjectionResult?.let { (resultCode, data) ->
            val mpManager = getSystemService(MEDIA_PROJECTION_SERVICE) as MediaProjectionManager
            mediaProjection = mpManager.getMediaProjection(resultCode, data)
            Companion.mediaProjection = mediaProjection
        }

        // Start HTTP server
        if (httpServer == null) {
            httpServer = CuaHttpServer()
            try {
                httpServer?.start()
                isRunning = true
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        return START_STICKY
    }

    override fun onDestroy() {
        httpServer?.stop()
        virtualDisplay?.release()
        imageReader?.close()
        mediaProjection?.stop()
        isRunning = false
        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? = null

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                "cua_service",
                "CUA 服务",
                NotificationManager.IMPORTANCE_LOW
            ).apply {
                description = "Hermes CUA 后台服务"
            }
            val nm = getSystemService(NotificationManager::class.java)
            nm.createNotificationChannel(channel)
        }
    }

    private fun createNotification(): Notification {
        val builder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Notification.Builder(this, "cua_service")
        } else {
            Notification.Builder(this)
        }
        return builder
            .setContentTitle("Hermes CUA 助手")
            .setContentText("运行中 · 端口 $PORT")
            .setSmallIcon(android.R.drawable.ic_menu_camera)
            .setOngoing(true)
            .build()
    }

    fun takeScreenshot(): ByteArray? {
        try {
            val wm = getSystemService(WINDOW_SERVICE) as WindowManager
            val display = wm.defaultDisplay
            val metrics = DisplayMetrics()
            display.getMetrics(metrics)

            val width = metrics.widthPixels
            val height = metrics.heightPixels
            val densityDpi = metrics.densityDpi

            val mp = mediaProjection ?: return null

            val format = android.graphics.ImageFormat.YUV_420_888
            imageReader?.close()
            imageReader = ImageReader.newInstance(width, height, format, 2)

            virtualDisplay?.release()
            virtualDisplay = mp.createVirtualDisplay(
                "cua-screenshot",
                width, height, densityDpi,
                DisplayManager.VIRTUAL_DISPLAY_FLAG_AUTO_MIRROR,
                imageReader?.surface, null, null
            )

            // Wait for image
            Thread.sleep(500)

            val image = imageReader?.acquireLatestImage() ?: return null
            val planes = image.planes
            if (planes.isEmpty()) return null

            // Convert YUV_420_888 to JPEG
            val buffer = planes[0].buffer
            val bytes = ByteArray(buffer.remaining())
            buffer.get(bytes)

            // For now, return raw YUV data
            // Proper JPEG conversion would use RenderScript or a library
            image.close()

            // Alternative: Use MediaProjection screenshot API (Android 14+)
            return captureScreenDirectly(metrics)
        } catch (e: Exception) {
            e.printStackTrace()
            return null
        }
    }

    private fun captureScreenDirectly(metrics: DisplayMetrics): ByteArray? {
        // Use the standard Android screenshot method
        // On Android 14+, we can use the HardwareBuffer based approach
        // For broader compatibility, we attempt a simpler approach
        return null
    }

    inner class CuaHttpServer : NanoHTTPD(PORT) {
        override fun serve(session: IHTTPSession): Response {
            val uri = session.uri
            val method = session.method

            return try {
                when {
                    uri == "/status" -> jsonResponse("""{"status":"running","port":$PORT,"accessibility":${CuaAccessibilityService.instance != null}}""")

                    uri == "/screenshot" -> handleScreenshot()

                    uri == "/tap" && method == Method.POST -> handleTap(session)
                    uri == "/swipe" && method == Method.POST -> handleSwipe(session)
                    uri == "/text" && method == Method.POST -> handleText(session)
                    uri == "/key" && method == Method.POST -> handleKey(session)

                    uri == "/ui" -> handleUiDump()

                    uri == "/info" -> jsonResponse(getDeviceInfo())

                    else -> newFixedLengthResponse(Response.Status.NOT_FOUND, "text/plain", "404")
                }
            } catch (e: Exception) {
                jsonResponse("""{"error":"${e.message}"}""", Response.Status.INTERNAL_ERROR)
            }
        }

        private fun handleScreenshot(): Response {
            // Use AccessibilityService screenshot (Android 14+)
            val accService = CuaAccessibilityService.instance
            if (accService != null) {
                accService.takeScreenshot()
                Thread.sleep(1000)
                val bytes = lastScreenshotBytes
                if (bytes != null) {
                    return newFixedLengthResponse(Response.Status.OK, "image/png", bytes)
                }
            }
            return jsonResponse("""{"error":"screenshot failed"}""", Response.Status.INTERNAL_ERROR)
        }

        private fun handleTap(session: IHTTPSession): Response {
            val params = parseBodyAndParams(session)
            val x = params["x"]?.toFloatOrNull() ?: return jsonResponse("""{"error":"missing x"}""", Response.Status.BAD_REQUEST)
            val y = params["y"]?.toFloatOrNull() ?: return jsonResponse("""{"error":"missing y"}""", Response.Status.BAD_REQUEST)

            CuaAccessibilityService.instance?.performTap(x, y)
            return jsonResponse("""{"status":"ok","x":$x,"y":$y}""")
        }

        private fun handleSwipe(session: IHTTPSession): Response {
            val params = parseBodyAndParams(session)
            val x1 = params["x1"]?.toFloatOrNull() ?: return jsonResponse("""{"error":"missing x1"}""")
            val y1 = params["y1"]?.toFloatOrNull() ?: return jsonResponse("""{"error":"missing y1"}""")
            val x2 = params["x2"]?.toFloatOrNull() ?: return jsonResponse("""{"error":"missing x2"}""")
            val y2 = params["y2"]?.toFloatOrNull() ?: return jsonResponse("""{"error":"missing y2"}""")

            CuaAccessibilityService.instance?.performSwipe(x1, y1, x2, y2)
            return jsonResponse("""{"status":"ok"}""")
        }

        private fun handleText(session: IHTTPSession): Response {
            val params = parseBodyAndParams(session)
            val text = params["text"] ?: return jsonResponse("""{"error":"missing text"}""")
            CuaAccessibilityService.instance?.performText(text)
            return jsonResponse("""{"status":"ok"}""")
        }

        private fun handleKey(session: IHTTPSession): Response {
            val params = parseBodyAndParams(session)
            val key = params["key"] ?: return jsonResponse("""{"error":"missing key"}""")

            val actionId = when (key.uppercase()) {
                "BACK" -> android.accessibilityservice.GestureDescription.GESTURE_GLOBAL_ACTION_BACK
                "HOME" -> android.accessibilityservice.GestureDescription.GESTURE_GLOBAL_ACTION_HOME
                "RECENTS" -> android.accessibilityservice.GestureDescription.GESTURE_GLOBAL_ACTION_RECENTS
                "NOTIFICATIONS" -> android.accessibilityservice.GestureDescription.GESTURE_GLOBAL_ACTION_NOTIFICATIONS
                else -> return jsonResponse("""{"error":"unknown key: $key"}""")
            }

            CuaAccessibilityService.instance?.performGlobalAction(actionId)
            return jsonResponse("""{"status":"ok","key":"$key"}""")
        }

        private fun handleUiDump(): Response {
            val accService = CuaAccessibilityService.instance
            if (accService == null) {
                return jsonResponse("""{"error":"accessibility service not connected"}""")
            }
            val uiXml = accService.getUiHierarchy()
            return newFixedLengthResponse(Response.Status.OK, "application/xml", uiXml ?: "<empty/>")
        }

        private fun parseBodyAndParams(session: IHTTPSession): Map<String, String> {
            val params = mutableMapOf<String, String>()
            // Parse query params
            session.parms?.let { params.putAll(it) }
            // Parse body as JSON
            try {
                val body = ByteArrayOutputStream()
                session.inputStream?.copyTo(body)
                val bodyStr = String(body.toByteArray())
                if (bodyStr.isNotBlank()) {
                    val json = com.google.gson.JsonParser.parseString(bodyStr).asJsonObject
                    json.entrySet().forEach { (key, value) ->
                        if (!params.containsKey(key)) {
                            params[key] = value.asString
                        }
                    }
                }
            } catch (_: Exception) {}
            return params
        }

        private fun jsonResponse(json: String, status: Response.Status = Response.Status.OK): Response {
            return newFixedLengthResponse(status, "application/json", json)
        }

        private fun getDeviceInfo(): String {
            val wm = getSystemService(WINDOW_SERVICE) as WindowManager
            val metrics = DisplayMetrics()
            wm.defaultDisplay.getMetrics(metrics)
            return """
            {
                "device":"${Build.MODEL}",
                "manufacturer":"${Build.MANUFACTURER}",
                "sdk":${Build.VERSION.SDK_INT},
                "release":"${Build.VERSION.RELEASE}",
                "width":${metrics.widthPixels},
                "height":${metrics.heightPixels},
                "density":${metrics.densityDpi}
            }
            """.trimIndent()
        }
    }
}
